File: Python/trex_selector_methods/trex/demo_trex_01_single_run/demo_trex_01_single_run.py
# ==============================================================================
# demo_trex_01_single_run.py
# ==============================================================================
#
# Classical T-Rex Selector demo — single-run.
# Mirrors cpp/trex_selector_methods/trex/demo_trex_01_single_run.cpp and
# R/trex_selector_methods/trex/demo_trex_01_single_run.R.
#
# Demo content:
#  - High- and low-dimensional settings (run in that order, as in C++).
#  - Fixed true support; fixed +1 coefficients (rnd_coef=False); TLARS solver.
#  - Reports selected indices, FDP/TPP, Phi_prime and Phi (printed sparsely),
#    FDP_hat, R, and the voting grid — dual console + file output to
#    simulation_results/data/d01_trex_basic_n{n}_p{p}.txt.
#
# Seed note: data seed 58 and selector seed 42 mirror the C++ demo
# structurally; the RNG streams differ across languages, so the drawn data
# (and hence the exact selection) are not bit-identical to the C++ output.
#
# ==============================================================================

import sys
import os

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

import numpy as np

import trex_selector_neo as tsn

from dgp_gauss_snr import dgp_gauss_snr
from trex_helpers import make_dual_out, print_single_run_result
from trex_sim_common import _make_trex_ctrl_from_dict

# ==============================================================================
# Global parameters
# ==============================================================================

# Fixed true support (0-based, same as C++ demo).
# C++ (0-based): {27, 149, 43, 128, 42, 4}
# R   (1-based): {28, 150, 44, 129, 43, 5}
TRUE_SUPPORT = np.array([27, 149, 43, 128, 42, 4], dtype=np.intp)

_OUT_DIR = os.path.join(_THIS_DIR, "simulation_results", "data")


# ==============================================================================
# Part 1: single-run basic demo
# ==============================================================================

def part_1_single_run():
    tFDR      = 0.1
    snr       = 1.0
    data_seed = 58   # mirrors C++ SyntheticData(..., seed=58)
    trex_seed = 42   # mirrors C++ TRexSelector(..., 42, true)

    ctrl_dict = dict(
        solver_name               = "TLARS",
        K                         = 20,
        max_dummy_multiplier      = 10,
        use_max_T_stop            = True,
        lloop_strategy            = "STANDARD",
        tloop_stagnation_stop     = False,
        tloop_max_stagnant_steps  = 5,
        parallel_rnd_experiments  = False,   # mirrors C++ / R use_openmp=FALSE
        lambda2                   = 0.1,
        rho_afs                   = 0.3,
        ncgmp_variant             = 0,
    )

    def run_single(high_dim, rnd_coef=False):
        n         = 150 if high_dim else 5000
        p         = 300 if high_dim else 1000
        dim_label = "High-dimensional (p > n)" if high_dim else "Low-dimensional (n > p)"

        # The C++ rnd_coef path selects a fixed hardcoded vector that mimics
        # randomly drawn coefficients (it is NOT a random draw); main() runs
        # with rnd_coef=False, i.e. all coefficients equal 1.
        coefs = (np.array([-0.4, -0.25, -0.8, 1.1, 2.5, -1.2])
                 if rnd_coef else np.ones(len(TRUE_SUPPORT)))

        os.makedirs(_OUT_DIR, exist_ok=True)
        out_path = os.path.join(_OUT_DIR, f"d01_trex_basic_n{n}_p{p}.txt")

        with open(out_path, "w") as fh:
            out = make_dual_out(fh)

            out(f"\n{'=' * 70}\n")
            out(f"Part 1: Single-run basic T-Rex demo  |  {dim_label}\n")
            out(f"n = {n},  p = {p},  tFDR = {tFDR:.2f},  SNR = {snr:.1f}\n\n")

            out("Generating synthetic data ...\n")
            dat = dgp_gauss_snr(n, p, TRUE_SUPPORT, coefs=coefs, snr=snr,
                                seed=data_seed)

            out("Running T-Rex Selector (TLARS) ...\n\n")
            ctrl = _make_trex_ctrl_from_dict(ctrl_dict)
            sel  = tsn.TRexSelector(
                np.asfortranarray(dat["X"]),
                dat["y"],
                tFDR=tFDR,
                trex_control=ctrl,
                seed=trex_seed,
                verbose=True,
            )
            sel.select()

            print_single_run_result(
                f"Part 1 — i.i.d. Gaussian  [{dim_label}, TLARS]",
                dat, sel, tFDR, out=out,
            )

        print(f"[Info] Results saved to: {out_path}")

    # C++ main() runs the high-dimensional scenario first, then the
    # low-dimensional one; both with fixed +1 coefficients.
    run_single(high_dim=True)
    run_single(high_dim=False)


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    part_1_single_run()
    print("\ndemo_trex_01_single_run.py complete.")

File: cpp/trex_selector_methods/trex/demo_trex_01_single_run/demo_trex_01_single_run.cpp
// ==============================================================================
// demo_trex_01_single_run.cpp
// ==============================================================================
/**
 * @file demo_trex_01_single_run.cpp
 *
 * @brief Demonstration of classical T-Rex Selector.
 *
 * @details Shows basic usage of the classical T-Rex Selector for both low- and
 *          high-dimensional settings in a single-run scenario.
 */
// ==============================================================================

// std includes
#include <fstream>
#include <iostream>
#include <iomanip>
#include <sstream>
#include <vector>

// Eigen includes
#include <Eigen/Dense>

// OpenMP compatibility layer
#include <utils/openmp/utils_openmp.hpp>

// T-Rex Selector includes
#include <trex_selector_methods/trex_core/trex.hpp>
#include <utils/datageneration/utils_datagen.hpp>
#include <utils/eval_metrics/utils_eval_cdiagnostics.hpp>
#include <utils/eval_metrics/utils_eval_rates.hpp>


// ==============================================================================
// Namespace aliases
// ==============================================================================

namespace cdianostics = trex::utils::eval::cdiagnostics;
namespace datagen = trex::utils::datageneration::datagen;
namespace dummygen= trex::utils::datageneration::dummygen;
namespace rates = trex::utils::eval::rates;


// T-Rex types
using trex::trex_selector_methods::trex_core::TRexSelector;
using trex::trex_selector_methods::trex_core::TRexControlParameter;
using trex::trex_selector_methods::trex_core::LLoopStrategy;
using trex::trex_selector_methods::utils::solver_dispatch::SolverTypeForTRex;

// ==============================================================================

// ==============================================================================
// Demo Single-Run: Basic T-Rex Selector functionality
// ==============================================================================

void demo_TRexSelector(bool high_dim, bool rnd_coef) {

    // Print Demo Header
    cdianostics::print_section_header("Demo: T-Rex Selector Basic Functionality");

    // Setup
    const Eigen::Index n = high_dim ? 150: 5000;
    const Eigen::Index p = high_dim ? 300 : 1000;

    const std::vector<std::size_t> true_support = {
        27, 149, 43 , 128, 42, 4
    };
    const std::vector<double> true_coefs = rnd_coef ?
        std::vector<double>{-0.4, -0.25, -0.8, 1.1, 2.5, -1.2} :
        std::vector<double>{1, 1, 1, 1, 1, 1};
    const double snr = 1.0;
    const double tFDR = 0.1;

    // Setup dual output (console + file)
    const std::string folder = DEMO_OUTPUT_DIR;
    std::filesystem::create_directories(folder);
    const std::string filename = "d01_trex_basic_n" + std::to_string(n) +
                                 "_p" + std::to_string(p) + ".txt";
    std::ofstream out_file(folder + filename);
    auto print_dual = [&](const std::string& text) {
        std::cout << text;
        if (out_file.is_open()) out_file << text;
    };

    print_dual(std::string(high_dim ? "High-dimensional (p > n)"
                                    : "Low-dimensional (n > p)") + "\n");

    // Generate synthetic data
    print_dual("Generating synthetic data...\n");
    datagen::SyntheticData data(n, p, true_support, true_coefs, snr, /*seed=*/58);

    // Create mapped views as lvalues (required by TRexSelector constructor)
    print_dual("Creating maps of data...\n");
    Eigen::Map<Eigen::MatrixXd> X_map(data.getX().data(),
                                      data.rows(),
                                      data.cols());
    Eigen::Map<Eigen::VectorXd> y_map(data.getY().data(),
                                      data.rows());

    // Setup Control Structures
    TRexControlParameter trex_ctrl;
    trex_ctrl.K = 20;
    trex_ctrl.max_dummy_multiplier = 10;
    trex_ctrl.use_max_T_stop = true;
    trex_ctrl.dummy_distribution = dummygen::Distribution::Normal();
    trex_ctrl.lloop_strategy = LLoopStrategy::STANDARD;
    trex_ctrl.tloop_stagnation_stop = false;
    trex_ctrl.tloop_max_stagnant_steps = 5;
    trex_ctrl.parallel_rnd_experiments = false;
    trex_ctrl.solver_type = SolverTypeForTRex::TLARS;

    // Create T-Rex Selector instance
    print_dual("Creating T-Rex Selector instance...\n");
    TRexSelector trex(X_map, y_map, tFDR, trex_ctrl, 42, true);

    // Execute T-Rex Selector
    print_dual("Executing T-Rex Selector...\n");
    trex.select();

    const auto selected_indices = trex.getSelectedIndices();
    {
        std::ostringstream ss;
        ss << "Selected indices: ";
        for (const auto& idx : selected_indices) ss << idx << " ";
        ss << "\n";
        print_dual(ss.str());
    }

    const double fdp = rates::compute_fdp(selected_indices, true_support);
    const double tpp = rates::compute_tpp(selected_indices, true_support);
    {
        std::ostringstream ss;
        ss << std::fixed << std::setprecision(4);
        ss << "False Discovery Proportion (FDP): " << fdp << "\n";
        ss << "True Positive Proportion (TPP):   " << tpp << "\n";
        print_dual(ss.str());
    }

    {
        const auto& phi_prime = trex.getPhiPrime();
        std::ostringstream ss;
        ss << "\nAdjusted Relative Occurrences (Phi_prime):\n"
           << phi_prime.transpose() << "\n";
        print_dual(ss.str());
    }

    {
        const auto& phi_mat = trex.getPhiMat();
        std::ostringstream ss;
        ss << "\nPhi Matrix (Phi):\n" << phi_mat << "\n";
        print_dual(ss.str());
    }

    {
        const auto& fdp_hat_mat = trex.getFDPHatMat();
        std::ostringstream ss;
        ss << "\nEstimated FDP Matrix (FDP_hat):\n" << fdp_hat_mat << "\n";
        print_dual(ss.str());
    }

    {
        const auto& r_mat = trex.getRMat();
        std::ostringstream ss;
        ss << "\nR Matrix (R):\n" << r_mat << "\n";
        print_dual(ss.str());
    }

    {
        const auto& voting_grid = trex.getVotingGrid();
        std::ostringstream ss;
        ss << "\nVoting Grid:\n" << voting_grid.transpose() << "\n";
        print_dual(ss.str());
    }

    print_dual("\n\n");

    if (out_file.is_open()) {
        std::cout << "[Info] Results saved to: " << folder + filename << "\n";
        out_file.close();
    }
}

// ==============================================================================
// Main
// ==============================================================================

int main() {

    std::cout.setf(std::ios::unitbuf);

    // Run basic T-Rex Selector demo without Monte Carlo simulation
    // --------------------------------------------------------------------------------------
    if (true) demo_TRexSelector(/*high_dim=*/true, /*rnd_coef=*/false);

    if (true) demo_TRexSelector(/*high_dim=*/false, /*rnd_coef=*/false);

    return 0;
}
